dataframe-arrow/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowWriter.kt

```diff
@@ -0,0 +1,491 @@
+//            is StructVector -> todo
+            else -> {
+                TODO("Saving to ${vector.javaClass.canonicalName} is not implemented")
+            }
+        }
+
+        vector.valueCount = dataFrame.rowsCount()
+    }
+
+    /**
+     * Create Arrow FieldVector with [column] content cast to [field] type according to [strictType] and [strictNullable] settings.
+     */
+    private fun allocateVectorAndInfill(field: Field, column: AnyCol?, strictType: Boolean, strictNullable: Boolean): FieldVector {
+        val containNulls = (column == null || column.hasNulls())
+        // Convert the column to type specified in field. (If we already have target type, convertTo will do nothing)
+
+        fun handleConversionFail(e: Exception): Pair<AnyCol?, Field> {
+            if (strictType) {
+                // If conversion failed but strictType is enabled, throw the exception
+                throw e
+            } else {
+                // If strictType is not enabled, use original data with its type. Target nullable is saved at this step.
+                warningSubscriber(e.message!!)
+                val actualType = listOf(column!!).toArrowSchema(warningSubscriber).fields.first().fieldType.type
+                val actualField = Field(field.name, FieldType(field.isNullable, actualType, field.fieldType.dictionary), field.children)
+                return column to actualField
+            }
+        }
+
+        val (convertedColumn, actualField) = try {
+            convertColumnToTarget(column, field.type) to field
+        } catch (e: TypeConversionException) {
+            handleConversionFail(e)
+        } catch (e: TypeConverterNotFoundException) {
+            handleConversionFail(e)
+        }
+
+        val vector = if (!actualField.isNullable && containNulls) {
+            if (strictNullable) {
+                throw IllegalArgumentException("Column \"${actualField.name}\" contains nulls but should be not nullable")
+            } else {
+                warningSubscriber("Column \"${actualField.name}\" contains nulls but expected not nullable")
+                Field(actualField.name, FieldType(true, actualField.fieldType.type, actualField.fieldType.dictionary), actualField.children).createVector(allocator)!!
+            }
+        } else {
+            actualField.createVector(allocator)!!
+        }
+
+        allocateVector(vector, dataFrame.rowsCount())
+        if (convertedColumn == null) {
+            check(actualField.isNullable)
+            infillWithNulls(vector, dataFrame.rowsCount())
+        } else {
+            infillVector(vector, convertedColumn)
+        }
+        return vector
+    }
+
+    private fun List<AnyCol>.toVectors(): List<FieldVector> = this.toArrowSchema(warningSubscriber).fields.mapIndexed { i, field ->
+        allocateVectorAndInfill(field, this[i], true, true)
```

Comment:
Maybe convert AnyCol to Field one by one, so you can get rid of this nullability in `this[i]`?

Reply:
like 
```
map { 
   val field = it.toArrowField() 
   allocateVectorAndInfill(field, it, true, true)
}
```

Reply:
Accepted, thanks